File: src/Autorep.Web/Client/faults/faultAggregator.ts
// TypeScript mirror of Domain/Faults/FaultAggregator.cs. Pure: a list of faults -> a FaultSummary
// grouped by component (each group rated by its worst fault, worst-severity-first). Pinned by the
// shared fixtures in tests/fixtures/faults (the same JSON the .NET xUnit suite uses).
import type { FaultSeverity } from "../wizard/types";

export interface FaultInput {
  /** Stable key (the source visual-item / reading key) — used to attach recommendations. */
  key?: string;
  component: string;
  description: string;
  severity: FaultSeverity;
  source: string;
  recommendation?: string;
}

export interface FaultGroup {
  component: string;
  severity: FaultSeverity;
  faults: FaultInput[];
}

export interface FaultSummary {
  groups: FaultGroup[];
  critical: number;
  major: number;
  minor: number;
  total: number;
}

const RANK: Record<FaultSeverity, number> = { Critical: 3, Major: 2, Minor: 1 };

export function aggregate(faults: FaultInput[]): FaultSummary {
  const byComponent = new Map<string, FaultInput[]>();
  for (const f of faults) {
    const list = byComponent.get(f.component) ?? [];
    list.push(f);
    byComponent.set(f.component, list);
  }

  const groups: FaultGroup[] = [...byComponent.entries()].map(([component, list]) => ({
    component,
    severity: list.reduce<FaultSeverity>((worst, f) => (RANK[f.severity] > RANK[worst] ? f.severity : worst), "Minor"),
    faults: list,
  }));
  groups.sort((a, b) => RANK[b.severity] - RANK[a.severity] || (a.component < b.component ? -1 : a.component > b.component ? 1 : 0));

  const count = (s: FaultSeverity) => faults.filter((f) => f.severity === s).length;
  return { groups, critical: count("Critical"), major: count("Major"), minor: count("Minor"), total: faults.length };
}

File: src/Autorep.Web/Domain/Faults/FaultAggregator.cs
namespace Autorep.Web.Domain.Faults;

public sealed record FaultInput(
    string Component, string Description, string Severity, string Source, string? Recommendation = null);

public sealed record FaultGroup(string Component, string Severity, IReadOnlyList<FaultInput> Faults);

public sealed record FaultSummary(IReadOnlyList<FaultGroup> Groups, int Critical, int Major, int Minor, int Total);

/// <summary>
/// Pure function: the test's faults (Visual Faults + failed numerical readings) → a FaultSummary
/// grouped by component, each group rated by its worst fault, ordered worst-severity-first. Mirrored
/// in TypeScript for the offline wizard; pinned by the shared fixtures in <c>tests/fixtures/faults</c>.
/// </summary>
public static class FaultAggregator
{
    private static int Rank(string severity) => severity switch
    {
        "Critical" => 3,
        "Major" => 2,
        _ => 1,
    };

    public static FaultSummary Aggregate(IEnumerable<FaultInput> faults)
    {
        var all = faults.ToList();

        var groups = all
            .GroupBy(f => f.Component)
            .Select(g => new FaultGroup(
                g.Key,
                g.Select(f => f.Severity).OrderByDescending(Rank).First(),
                g.ToList()))
            .OrderByDescending(g => Rank(g.Severity))
            .ThenBy(g => g.Component, StringComparer.Ordinal)
            .ToList();

        int Count(string s) => all.Count(f => f.Severity == s);
        return new FaultSummary(groups, Count("Critical"), Count("Major"), Count("Minor"), all.Count);
    }
}
